packages/core/src/lib/passkey/options.ts

```diff
@@ -0,0 +1,130 @@
+  )
+
+  // Get the full user from the email
+  const user = email ? await adapter.getUserByEmail(email) : null
+
+  // Find the user's account associated with the provider
+  const accounts = user ? (await adapter.listLinkedAccounts(user.id)) ?? [] : []
+  const account = accounts.find((a) => a.provider === provider.id) ?? null
+
+  // Find the account's authenticators
+  const authenticators = account
+    ? (await adapter.listAuthenticatorsByAccountId(
+      account.providerAccountId
+    )) ?? undefined
+    : undefined
+
+  return [authenticators, user, account]
+}
+
+/**
+ * Generate passkey authentication options.
+ * If a user is provided, their credentials will be used to generate the options.
+ * Otherwise, allow any credentials.
+ *
+ * @param options
+ * @param email Optional user email to use to generate the options.
+ * @returns The options prepared for the client.
+ */
+export async function authenticationOptions(
+  options: Options,
+  email?: string
+): Promise<PublicKeyCredentialRequestOptionsJSON> {
+  const { provider } = options
+
+  // Get the user's authenticators
+  const [authenticators] = await getUserAndAuthenticators(options, email)
+
+  // Generate authentication options
+  const authOptions = await generateAuthenticationOptions({
+    rpID: provider.relayingParty.id,
+    allowCredentials: authenticators?.map((a) => ({
+      id: a.credentialID,
+      type: "public-key",
+      transports: a.transports,
+    })),
+    userVerification: "preferred",
+  })
+
+  return authOptions
+}
+
+/**
+ * Generate passkey registration options.
+ * If a user is provided, their credentials will be used to generate the options.
+ * Otherwise, their email will be used to generate the options.
+ *
+ * @param options
+ * @param email The user's email to use to generate the options.
+ * @returns The options prepared for the client.
+ */
+export async function registrationOptions(
+  options: Options,
+  email: string
+): Promise<PublicKeyCredentialCreationOptionsJSON> {
+  const { provider } = options
+
+  // Get the user authenticators and user object
+  const [authenticators, user, account] = await getUserAndAuthenticators(options, email)
+
+  // Generate a random acc ID and user name if the user does not exist
+  const accountProviderID = account?.providerAccountId ?? randomString(32)
+  const userName = user?.name ?? user?.email ?? email
+  const userDisplayName = user?.name ?? userName
+
+  // Generate registration options
+  const regOptions = await generateRegistrationOptions({
+    userID: accountProviderID,
+    userName,
+    userDisplayName,
+    rpID: provider.relayingParty.id,
+    rpName: provider.relayingParty.name,
+    excludeCredentials: authenticators?.map((a) => ({
+      id: a.credentialID,
+      type: "public-key",
+      transports: a.transports,
+    })),
+    authenticatorSelection: {
+      residentKey: "preferred",
+      userVerification: "preferred",
+      requireResidentKey: true,
```

Comment:
In the [spec](https://w3c.github.io/webauthn/#dictionary-authenticatorSelection), it is said that this should only be set to `true` if `residentKey` is `required`
> [Relying Parties](https://w3c.github.io/webauthn/#relying-party) SHOULD set it to true if, and only if, [residentKey](https://w3c.github.io/webauthn/#dom-authenticatorselectioncriteria-residentkey) is set to [required](https://w3c.github.io/webauthn/#dom-residentkeyrequirement-required).

since`residentKey` is being set to `preferred`, I think we can remove this value. 👀

Reply:
You're correct, but my mistake here was not setting `residentKey` to `required`.
I'll remove `requireResidentKey` since it only exists to support the previous spec.

We wan't `required` here because passkeys are always "resident keys" as [Discoverable Credentials](https://passkeys.dev/docs/reference/terms/#discoverable-credential)

Fixed in 2a224e30e67fa041f8093123364d233aaaf62351